lib/ui/semantics.dart

```diff
@@ -641,6 +641,121 @@ class SemanticsFlag {
   }
 }
 
+/// A string that carries a list of [StringAttribute]s.
+class AttributedString extends NativeFieldWrapperClass2 {
+  /// Creates a attributed string.
+  ///
+  /// The [TextRange] in the [attributes] must be inside the length of the
+  /// [string].
+  AttributedString(
+    String string, {
+    List<StringAttribute>? attributes,
+  }) : assert((){
+         if (attributes != null) {
+           for (final StringAttribute attribute in attributes) {
+             assert(
+               string.length >= attribute.range.start &&
+               string.length >= attribute.range.end,
+               'The range in $attribute is outside of the string $string',
+             );
+           }
+         }
+         return true;
+       }()) {
+    _constructor();
+    _initAttributedString(string, attributes);
+  }
+
+  AttributedString._withConcat(
+    AttributedString first,
+    AttributedString second
+  ) {
+    _constructor();
+    _initWithConcat(first, second);
+  }
+
+  /// The plain string stored in the attributed string.
+  String get string => _getString();
```

Comment:
I'm not clear on why we would require a native hop to do this. I'm also not clear on why we would expose this but not the attributes.

Why don't we just make a `final String _string` field, and return that here as well as passing it into the native methods that need it?

Reply:
In particular, it'll be relatively expensive to repeatedly access this getter, even though it seems like it shouldn't be, since it will require a hop over to native and a copy of the string back into a dart object.

Reply:
I did this because of the concat method, for any string operations(join, substring...), I will have to duplicate code on both dart and native side if i cache the string. attribute will be even more complicated if we expose it. there is also a question what should we do if user mutate the attributes if we expose it, I remember we decide to not expose it for now and see if there is any use case that user wants to see what they have set for the attribute.

Reply:
IMHO, we should either not expose the string at all or we should have it stored on the Dart side. And if we store both, it should make concat easier to just implement solely on the Dart side right?

Reply:
ah you are right, we can cancate the string in dart side and call the default constructor. I will do that